src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,61 @@
+use crate::{hash, hash_pair};
+
+pub struct MerkleTree {
+    hashes: Vec<Vec<u8>>,
+    count: usize
+}
+
+impl MerkleTree {
+    pub fn new(elements: Vec<String>) -> MerkleTree{
+        let leaves = Self::hash_elements(elements.clone());
+        let hashes = Self::build_hashes(leaves);
+        MerkleTree{hashes, count: elements.len()}
+    }
+
+    pub fn root_hash(tree: MerkleTree) -> Vec<u8> {
+        tree.hashes[0].clone()
+    }
+
+    pub fn add(tree: MerkleTree, elements: Vec<String>) -> MerkleTree {
+        let new_leaves = Self::hash_elements(elements);
+        let old_leaves = tree.hashes[{tree.count-1}..].to_vec();
+        let leaves = [old_leaves, new_leaves].concat();
+        let hashes = Self::build_hashes(leaves.clone());
+        MerkleTree{hashes, count: leaves.len()}
+    }
+
+    fn hash_elements(elements: Vec<String>) -> Vec<Vec<u8>> {
+        elements.iter().map(|e| hash(e.to_string())).collect()
+    }
+
+    fn build_hashes(hashes: Vec<Vec<u8>>) -> Vec<Vec<u8>> {
+        if hashes.len() == 1 {
+            return hashes;
+        }
+
+        let h: Vec<Vec<u8>>= hashes.chunks(2).map(|e| hash_pair(e[0].clone(), e[1].clone())).collect();
```

Comment:
i think that if you use chunks(2).clone().into_iter, you can avoid to clone the hashes
```suggestion
        let h: Vec<Vec<u8>>= hashes.chunks(2).clone().into_iter().map(|e| hash_pair(e[0], e[1])).collect();
```
not completely sure but i think that yes, mainly because you are using `map` directly into a `vec`, vec structure is iterable by default and when you call a higher order function into it, it will use the `iter()` method under the hood. This method returns references to the elements of the vector thus forcing you to use `.clone()` to respect ownership, instead `into_iter()` return the actual values of the vector and you may work upon them but you consume the vector in the process, that is why you clone it before pass it to `into_iter()`.

Reply:
The compiler still wants me to add .clone() to the elements 😪